api/websocket/messagebuffer/messagebuffer.go

```diff
@@ -1,39 +1,218 @@
 package messagebuffer
 
 import (
-	"encoding/hex"
+	"fmt"
 	"sync"
+	"time"
 
+	"github.com/nknorg/nkn/v2/config"
 	"github.com/nknorg/nkn/v2/pb"
 )
 
-// MessageBuffer is the buffer to hold message for clients not online
-type MessageBuffer struct {
-	sync.Mutex
-	buffer map[string][]*pb.Relay
+type Item struct {
+	relay        *pb.Relay
+	creationTime time.Time
+	expiryTime   time.Time
 }
 
-// NewMessageBuffer creates a MessageBuffer
-func NewMessageBuffer() *MessageBuffer {
-	return &MessageBuffer{
-		buffer: make(map[string][]*pb.Relay),
+type queue struct {
+	parent *Buffer
+	item   *Item
+}
+
+type Buffer struct {
+	mux   sync.Mutex
+	queue map[string][]*queue
+	size  int
+}
+
+func New() *Buffer {
+	return &Buffer{
+		queue: make(map[string][]*queue, 0),
+	}
+}
+
+func getParent(index int) int {
+	if index <= 0 {
+		return -1
+	}
+
+	return (index - 1) / 2
+}
+
+func getChildren(index int) (int, int) {
+	tree := index * 2
+
+	return tree + 1, tree + 2
+}
+
+func (q *queue) isAfter(node *queue) bool {
+	a := q.item
+	b := node.item
+
+	if a.creationTime.Before(b.creationTime) {
+		return false
+	}
+	if a.creationTime == b.creationTime {
+		return false
+	}
+
+	return true
+}
+
+func (q *queue) isAfterOrEqual(node *queue) bool {
+	a := q.item
+	b := node.item
+
+	if a.creationTime.Before(b.creationTime) {
+		return false
+	}
+	if a.creationTime == b.creationTime {
+		return true
+	}
+
+	return true
+}
+
+func (m *Buffer) Purge() {
+	var clients []string
+
+	m.mux.Lock()
+	for clientIDStr := range m.queue {
+		clients = append(clients, clientIDStr)
+	}
+	m.mux.Unlock()
+
+	for _, clientIDStr := range clients {
+		relays := m.PopMessages(clientIDStr)
+
+		for _, relay := range relays {
+			m.put(clientIDStr, relay)
+		}
 	}
 }
 
-// AddMessage adds a message to message buffer
-func (messageBuffer *MessageBuffer) AddMessage(clientID []byte, msg *pb.Relay) {
-	clientIDStr := hex.EncodeToString(clientID)
-	messageBuffer.Lock()
-	defer messageBuffer.Unlock()
-	messageBuffer.buffer[clientIDStr] = append(messageBuffer.buffer[clientIDStr], msg)
+func (m *Buffer) AddMessage(clientID string, relay *pb.Relay) error {
+	if relay.MaxHoldingSeconds <= 0 {
+		return nil
+	}
+
+	m.Purge()
```

Comment:
Calling Purge in AddMessage have two unwanted results:
1. AddMessage has O(N) time complexity
2. If no new messages arrive, old messages will never be removed